RecoLocalTracker/SubCollectionProducers/src/TrackClusterRemoverPhase2.cc

```diff
@@ -0,0 +1,223 @@
+
+/* This is a copy of the TrackClusterRemover 
+ * for Phase2 Tk upgrade
+ * FIXME:: changing with new phase2 pixel DataFormats!
+ * FIXME:: still to be factorized
+ */
+
+namespace {
+
+  class TrackClusterRemoverPhase2 final : public edm::global::EDProducer<> {
+  public:
+    TrackClusterRemoverPhase2(const edm::ParameterSet& iConfig) ;
+    ~TrackClusterRemoverPhase2(){}
+    static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+  private:
+    
+    virtual void produce(edm::StreamID, edm::Event& evt, const edm::EventSetup&) const override;
+ 
+    using PixelMaskContainer    = edm::ContainerMask<edmNew::DetSetVector<SiPixelCluster>>;
+    using Phase2OTMaskContainer = edm::ContainerMask<edmNew::DetSetVector<Phase2TrackerCluster1D>>;
+
+    using QualityMaskCollection = std::vector<unsigned char>;
+
+    const float maxChi2_;
+    const int minNumberOfLayersWithMeasBeforeFiltering_;
+    const reco::TrackBase::TrackQuality trackQuality_;
+
+
+    const TrackCollectionTokens trajectories_;
+    edm::EDGetTokenT<QualityMaskCollection> srcQuals;
+    
+    const edm::EDGetTokenT<edmNew::DetSetVector<SiPixelCluster> > pixelClusters_;
+    const edm::EDGetTokenT<edmNew::DetSetVector<Phase2TrackerCluster1D> > phase2OTClusters_;
+    
+    edm::EDGetTokenT<PixelMaskContainer> oldPxlMaskToken_;
+    edm::EDGetTokenT<Phase2OTMaskContainer> oldPh2OTMaskToken_;
+    
+    // backward compatibility during transition period
+    edm::EDGetTokenT<edm::ValueMap<int>>  overrideTrkQuals_;
+    
+  };
+
+  void TrackClusterRemoverPhase2::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+    edm::ParameterSetDescription desc;
+    desc.add<edm::InputTag>("trajectories",edm::InputTag());
+    desc.add<edm::InputTag>("trackClassifier",edm::InputTag("","QualityMasks"));
+    desc.add<edm::InputTag>("phase2OTPixelClusters",edm::InputTag("siPixelClusters"));
+    desc.add<edm::InputTag>("phase2OTClusters",edm::InputTag("siPhase2Clusters"));
+    desc.add<edm::InputTag>("oldClusterRemovalInfo",edm::InputTag());
+
+    desc.add<std::string>("TrackQuality","highPurity");
+    desc.add<double>("maxChi2",30.);
+    desc.add<int>("minNumberOfLayersWithMeasBeforeFiltering",0);
+    // old mode     
+    desc.add<edm::InputTag>("overrideTrkQuals",edm::InputTag());
+
+    descriptions.add("phase2trackClusterRemover", desc);
+ 
+  }
+  
+  TrackClusterRemoverPhase2::TrackClusterRemoverPhase2(const edm::ParameterSet& iConfig) :
+    maxChi2_(iConfig.getParameter<double>("maxChi2")),
+    minNumberOfLayersWithMeasBeforeFiltering_(iConfig.getParameter<int>("minNumberOfLayersWithMeasBeforeFiltering")),
+    trackQuality_(reco::TrackBase::qualityByName(iConfig.getParameter<std::string>("TrackQuality"))),
+
+    trajectories_(iConfig.getParameter<edm::InputTag>("trajectories"),consumesCollector()),
+    pixelClusters_(consumes<edmNew::DetSetVector<SiPixelCluster> >(iConfig.getParameter<edm::InputTag>("phase2OTPixelClusters"))),
```

Comment:
So, what I would propose is something like: phase2Pixel, phase2OTpixel and phase2OTstrips. OT stands for "Outer Tracker". Otherwise can be something like phase2ITPixel, phase2OTpixel and phase2OTstrips. The point is just that with the phase2 tracker, the two sub-detector have always been called "pixel" and "outer tracker" to define one or the other, so I think it is convenient to keep this nomenclature, but we can also think about introducing a "inner tracker" if this can help in the code.